Add lecture 3

joelostblom opened the pull request:
@QuLogic @lcoome I am involving more people in the reviews, so that @lwjohnst86 doesn't get overloaded =)

joelostblom commented:
This is for Thursday's lecture this week.

joelostblom commented:
@lwjohnst86 I fixed the first Travis error, but not sure what's happening in the slides generation on this second one:

```
/usr/bin/pandoc +RTS -K512m -RTS lec03-basic-r.utf8.md --to html --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output /home/travis/build/UofTCoders/rcourse/_site/slides/lec03-basic-r.html --smart --email-obfuscation none --self-contained --variable transition=0.01 --template /home/travis/R/Library/rmarkdown/rmd/ioslides/default.html --include-in-header /tmp/RtmpP5vdOe/rmarkdown-str29be6cd095f8.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' 

Killed

Error: pandoc document conversion failed with error 137

Execution halted

The command "Rscript _commands/build-slides.R" exited with 1.

0.01s$ touch _site/.nojekyll
```

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+> * Add new columns to a data frame that are functions of existing columns with `mutate`.
+> * Understand the split-apply-combine concept for data analysis.
+> * Use `summarize`, `group_by`, and `tally` to split a data frame into groups of observations, apply a summary statistics for each group, and then combine the results.
+
```
Don't forget to include the `Lesson Outline` with approximate times. It's better to overestimate time by a bit then to underestimate. I'll do a more thorough review later today.

joelostblom replied:
Yes, I will include this now!

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
 ---
 
+
+**Copyright (c) Data Carpentry**
```
Are you going to put this at the top of every lecture? You don't need to. Just put a note in the resources file and as a note at the end of this lecture.

If you are going to talk about it again, I'd suggest removing this line here, since this material is *not* copyrighted by Data Carpentry, but by us. (Of which we are using some of the material)

joelostblom replied:
Yes, I was originally planning to always include the copyright (but not talk about it in length ever lecture), because that is how I interpret [their licence for this material](http://www.datacarpentry.org/R-ecology-lesson/LICENSE.html):

> You are free: 
> to Share—to copy, distribute and transmit the work
> to Remix—to adapt the work
>
> Under the following conditions:
>    Attribution—You must attribute the work using “Copyright (c) Data Carpentry” ...

But I also agree with you that now we (also?) have copyright since we have modified the material. Maybe we can write that the original material is copyright by Data Carpentry and the modified is copyright by UofT Coders? Technically, I do not know if that is correct or if Data Carpentry also has copyright on our modified version. Do you know the details of cases like these?

Maybe @QuLogic can weigh in also.

lwjohnst86 replied:
They are licensed under the CC-BY, so as long as you indicate somewhere that some of the material was taken from them, it is fine. That's the beauty of CC-BY! So just put at the very bottom something like:

```
Parts of this lesson material were taken and modified from [Data Carpentry](http://datacarpentry.org) under their CC-BY copyright license. See their [lesson page](link/to/lesson/) for the original source.
```

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+> * Select certain columns in a data frame with the **`dplyr`** function `select`.
+> * Select certain rows in a data frame according to filtering conditions with the **`dplyr`** function `filter` .
+> * Link the output of one **`dplyr`** function to the input of another function with the 'pipe' operator `%>%`.
+> * Add new columns to a data frame that are functions of existing columns with `mutate`.
```
Lines 24 to 27 could be combined into saying "Learn to use data wrangling commands `select`, `filter`, `%>%`, and `mutate` from the `dplyr` package."

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+> * Use `summarize`, `group_by`, and `tally` to split a data frame into groups of observations, apply a summary statistics for each group, and then combine the results.
+
+
+## Data set background
```
`Dataset`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+
+Today, we will be working with real data from a longitudinal study of the
+species abundance in the Chihuahuan desert ecosystem near Portal, Arizona, USA.
+This study includes observations of plants, ants and rodents from 1977 - 2002,
```
Oxford comma "ants, and rodents"

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+[the abstract of this paper from 2009](
+http://onlinelibrary.wiley.com/doi/10.1890/08-1222.1/full). There are several
+data sets available related to this study, and we will be working with data sets that
+have been preprocessed by the data carpentry organization to facilitate
```
`by [Data Carpentry](https://www.datacarpentry.org) to`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+http://onlinelibrary.wiley.com/doi/10.1890/08-1222.1/full). There are several
+data sets available related to this study, and we will be working with data sets that
+have been preprocessed by the data carpentry organization to facilitate
+teaching. These are made available online as _The Portal project teaching
```
`The Portal Project Teaching Database`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+database_, both at the [data carpentry
+website](http://www.datacarpentry.org/ecology-workshop/data/), and on
+[Figshare](https://figshare.com/articles/Portal_Project_Teaching_Database/1314459/6),
+which is a great place to publish data and figures openly to make them available
```
`publish data, code, figures, and more`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+paper.
+
+
+## Presentation of the Survey Data
```
`survey data`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+
+| Column           | Description                        |
+|------------------|------------------------------------|
+| record_id       | Unique id for the observation      |
```
either each row is capitalized or they all are lower case. (`Unique` vs `month` right below)

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+| weight           | weight of the animal in grams      |
+| genus            | genus of animal                    |
+| species          | species of animal                  |
+| taxa             | e.g. Rodent, Reptile, Bird, Rabbit |
```
lowercase

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+| plot_type       | type of plot                       |
+
+To read the data into R, we are going to use a function called `read_csv`. This
+function is contained in an R-package called `readr`. R-packages are a bit like
```
Might want to link to the readr website. http://readr.tidyverse.org/

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+function is contained in an R-package called `readr`. R-packages are a bit like
+browser extensions, they are not essential, but can provide nifty functionality.
+We will go through R-packages in general and which ones are good for data
+analyses in particular in detail later in this lecture. Now, let's install
```
`analyses in detail later`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+```
+
+Now we can use the `read_csv` function. One useful thing that `read_csv`
+enables, is the ability to read a CSV file directly from a URL, without
````
`enables is the`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+```
+
+However, it is often a good idea to download the data first, so you have a copy
+stored locally on your computer in case you want to do some offline analyses or
````
`offline analyses,`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+
+However, it is often a good idea to download the data first, so you have a copy
+stored locally on your computer in case you want to do some offline analyses or
+the file online changes or is taken down. You can either download the data
```
`online changes, or the file`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+However, it is often a good idea to download the data first, so you have a copy
+stored locally on your computer in case you want to do some offline analyses or
+the file online changes or is taken down. You can either download the data
+manually or form within R:
```
`from within`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+surveys
+```
+
+You can see that the R Notebook provides us with this nice tabular display
````
On the website, it won't look like in the Notebook, so this sentence won't make sense for those reading the website. I would say this in person rather than in the material

joelostblom replied:
I will probably add specification of how it is looking in the knitted output and contrast it with what it would look like in the R Notebook.

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+`tibble` rather than a data frame as indicated in the output. This is because we
+used `read_csv` to load it. Since a `tibble` is just a data frame with some
+convenient extra functionality, we will use these words interchangeably from now
+on.
```
You might have to reword this paragraph when the first sentence is removed/rephrased.

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+use for statistics and plotting.
+
+A data frame can be created by hand, but most commonly they are generated by the
+function `read_csv()`in other words, when importing spreadsheets from your hard
```
```
`read_csv()`. In other words,
```

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+drive (or the web).
+
+A data frame is the representation of data in the format of a table where the
+columns are vectors that all have the same length. Because the column are
```
`Because the columns are`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+columns are vectors that all have the same length. Because the column are
+vectors, they all contain the same type of data as we discussed in last class
+(e.g., characters, integers, factors). For example, here is a figure depicting a
+data frame comprising of a numeric, a character and a logical vector.
```
Oxford comma. `a character, and a logical`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+treating them like strings.
+
+
+## Inspecting `data.frame` Objects
```
`objects`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+functions to get a sense of the content/structure of the data. Let's try them
+out!
+
+* Size:
```
I like this list :)

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+    * `rownames(surveys)` - returns the row names
+
+* Summary:
+    * `str(surveys)` - structure of the object and information about the class, length and
```
oxford comma: `length, and`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+out!
+
+* Size:
+    * `dim(surveys)` - returns a vector with the number of rows in the first element,
```
remove last comma `first element and`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+Note: most of these functions are "generic", they can be used on other types of
+objects besides `data.frame`.
+
+> ### Challenge
```
Bold challenge (otherwise it will be added to toc) ... actually, it may be nice to include it in the toc. Is that your reasoning? We can update the last lecture if yes :)

lwjohnst86 replied:
It would be nice if the challenges were shown on screen for longer. Sometimes I would miss what you said about doing a challenge and I'm sure other students may have too. If it's on screen it's hard to miss.

joelostblom replied:
Good point! I will make sure that the challenge stays up there for longer and hopefully also find a way that I can start typing out what I need to go through the answers. I find that screen estate is always a problem when working with super zoomed in projector displays.

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+surveys[, 1]    # first column in the data frame (as a vector)
+surveys[1]      # first column in the data frame (as a data.frame)
+surveys[1:3, 7] # first three elements in the 7th column (as a vector)
+surveys[3, ]    # the 3rd element for all columns (as a data.frame)
```
You have to be careful here. All of these indexing return a `data.frame`/`tbl_df`, **not** a vector. That's because this is a `tibble` object and so will always return a `tibble` object *unless* you use `[[` indexing. The reason is because the tibble package is trying to remove some strange inconsistencies (as you describe).

joelostblom replied:
Ah, nice catch. I originally was working with standard data frames durign the first part of the lecture. When I switch to introducing tibbles right from the beginning, I forgot that it would impact this indexing since it consistently always returns a tibble as you said. Will mention this convenient consistency and remove the erroneous comments.

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+
+```{r}
+head(surveys["species_id"])       # Result is a data.frame
+head(surveys[, "species_id"])     # Result is a vector
````
This line is not correct. It results in a `data.frame`/`tbl_df`.

joelostblom replied:
Actually, this confused me a bit now that I went through it. I was expecting that I would get a `tbl_df` back every time, just as with the numerical indexing, but the two last indexing options actually returns a vector...

joelostblom replied:
Ah ok, I got part of it at least. I never included double brackets `[[]]` for the numerical indeces. Doing something like `species[[1]]` also returns a vector. I guess `$` just is shorthand for the `[[]]` type indexing since that also returns a vector?

joelostblom replied:
I think I will just leave out `[[]]` indexing... `$` I feel is common though, is that correct? I mean, with `dplyr` we use `select` anyways so I guess I can keep it simple here.

lwjohnst86 replied:
Exactly. `$` is pretty common so I would talk about it. And yes `$` is short for `[[` since using `$` you can only select one column that way.

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+head(surveys$species_id)          # Result is a vector
+```
+
+For our purposes, the last three notations are equivalent. RStudio knows about
````
`last two notations`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+the columns in your data frame, so you can take advantage of the autocompletion
+feature to get the full and correct column name.
+
+> ### Challenge
```
Make sure to give them enough time to finish these challenges. I would suggest including that time in your `Lesson outline` in the top of the file.

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+--->
+
+
+## Exploratory data analyses (EDA) in R
```
I don't like how this header and the follow `###` headers look in the TOC on the HTML version. Can you just say `## R packages for data analysis` and change the `###` headers to `##`?

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+
+### R packages
+
+First, let's explain the concept of an R-package. What we have used so far, is
```
`so far is`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+
+First, let's explain the concept of an R-package. What we have used so far, is
+all part of base R (except `read_csv`), together with many more functions. Every
+package included in base R, will be installed on any computer where R is
```
`base R will be`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+```
+
+
+### Data wrangling with dplyr
````
I'd suggest `## Selecting columns and filtering rows with dplyr`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+carpentry](http://www.datacarpentry.org/R-ecology-lesson/05-r-and-databases.html)).
+
+
+### Selecting columns and filtering rows
```
... then remove this line.

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+in order to understand it. The package **`dplyr`** provides easy tools for the
+most common data manipulation tasks. It is built to work directly with data
+frames, with many common tasks optimized by being written in a compiled language
+(C++), this means that many operations runs much faster than for similar tools
```
`operations run much faster than similar tools`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+is great because you are allowed to work with much bigger datasets (100s of GB)
+than your computer could normally handle. We will not talk in detail about this
+in class, but there are great resources online to learn more (e.g. [this lecture
+from data
```
`Data Carpentry`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+```
+
+
+### Pipes
````
Maybe something like `## Chaining functions together using the pipe`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+surveys %>% 
+    select(plot_id, species_id, weight, year) %>% 
+    filter(year == 1995)
+```
````
Might be good to explain why you don't need to include the first argument.

```
surveys %>% select(., plot_id)
```

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+  select(species_id, sex, weight)
+```
+
+In the above, we use the pipe to send the `surveys` dataset first through
````
`above code`

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+    select(sex, taxa)
+```
+
+```{r, eval=FALSE}
````
why `eval=FALSE`?

joelostblom replied:
Not sure....

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+--->
+
+
+### Mutate
```
`## Creating new colums with dplyr`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+--->
+
+
+### Split-apply-combine data analysis and the summarize() function
```
`## Split-apply-combine technique in dplyr`

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+Many data analysis tasks can be approached using the *split-apply-combine*
+paradigm: split the data into groups, apply some analysis to each group, and
+then combine the results. **`dplyr`** makes this very easy through the use of
+the `group_by()` function.
```
```
the `group_by()` and `summarize()` function.
```

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+the `group_by()` function.
+
+
+#### The `summarize()` function
```
I don't think this header is necessary

lwjohnst86 commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,732 @@
+When grouping both by `sex` and `species_id`, the first rows are for individuals
+that escaped before their sex could be determined and weighted. You may notice
+that the last column does not contain `NA` but `NaN` (which refers to "Not a
+Number", this is only visible in the interactive Notebook, not the knitted output). To avoid this, we can remove the missing values for weight before we
```
line wrap

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+    summarize(mean_weight = mean(weight))
+```
+
+Note that the number of not displayed rows decreased from 82 to 54. If you want
````
"number of not displayed"? Do you mean `number of displayed rows`?

lwjohnst86 commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,732 @@
+```
+
+
+#### Tallying
````
Maybe `## Using tally to summarize categorical data` or something a bit more informative.

lwjohnst86 replied:
Reason I suggest *not* using `###` is because you use `###` for challenges, which makes it nicely separated on the TOC, so it's easy to quickly scan for challenges.

joelostblom replied:
Will either change this or make challenges `####`. I like having the challenges in the TOC so will switch back the previous lecture.

lwjohnst86 replied:
Yes, I agree, Challenges in TOC are actually a good idea. Though I wouldn't recommend using `####` since it may not show up in the TOC without changing some options and a fourth level header just doesn't look good in the TOC.

joelostblom replied:
I would prefer the third level for the subsections as I have it right now, so I used fourth level for the challenges... It is not a big deal if they don't show up in the TOC, so up to you if you want to include 4th level headings or not.

joelostblom commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+columns are vectors that all have the same length. Because the columns are
+vectors, they all contain the same type of data as we discussed in last class
+(e.g., characters, integers, factors). We can see this when inspecting the
+**str**ucture of a data frame with the function `str()`:
```
For me it looks like all text from here is bold in the GitHub pull request. It looks good locally and works to knit.

QuLogic replied:
Inconsistent markdown specifications...

joelostblom replied:
I see, I removed the bold for now. Although it would render fine on the home page, I don't think it is necessary.

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+[`readr`](http://readr.tidyverse.org/). R-packages are a bit like browser
+extensions, they are not essential, but can provide nifty functionality. We will
+go through R-packages in general and which ones are good for data analyses in
+in detail later in this lecture. Now, let's install `readr`:
```
Double 'in'

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+tabular view of the data, which also includes pagination when there are many
+rows and we can click this arrow to view all the columns. Technically, this is
+actually a `tibble` rather than a data frame as indicated in the output. This is
+because we used `read_csv` to load it. Since a `tibble` is just a data frame
```
There are a lot of 'this's in the previous 3 sentences.

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+
+* Summary:
+    * `str(surveys)` - structure of the object and information about the class,
+    * length, and
```
This bullet point shouldn't be here; probably the newline isn't needed either.

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+surveys[1, 1]   # first element in the first column of the data frame
+surveys[1, 6]   # first element in the 6th column
+surveys[, 1]    # first column in the data frame
+surveys[1]      # first column in the data frame
```
Well, that's weird... I guess no comma triggers a different kind of indexing?

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+package included in base R will be installed on any computer where R is
+installed, since they are considered critical for using R, e.g. `c()`, `mean()`,
+`+`, `-`, etc. However, since R is an open language, it is easy to develop your
+own R-packages that provides new functionality and submit it to the official
```
R-package

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+
+We will be using this package a lot, and it would be a little annoying to have
+to type `dplyr::` every time, so we will load it into our current environment.
+This needs to be done once for every new R sessions and makes all functions
```
session

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+We will be using this package a lot, and it would be a little annoying to have
+to type `dplyr::` every time, so we will load it into our current environment.
+This needs to be done once for every new R sessions and makes all functions
+accessible without the their package prefix, which is very convenient, as long
```
Extra 'the'

QuLogic commented on lec03-basic-r.Rmd:
````diff
@@ -1,4 +1,748 @@
+the same name from two different packages.
+
+```{r}
+# We could also do library(dplyr), but we need the rest of the
````
Missing quotes.

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+in order to understand it. The package **`dplyr`** provides easy tools for the
+most common data manipulation tasks. It is built to work directly with data
+frames, with many common tasks optimized by being written in a compiled language
+(C++), this means that many operations runs much faster than similar tools
```
run

QuLogic commented on lec03-basic-r.Rmd:
```diff
@@ -1,4 +1,748 @@
+(C++), this means that many operations runs much faster than similar tools
+in R. An additional feature is the ability to work directly with data stored in
+an external database, such as SQL-databases. The ability to work with databases
+is great because you are allowed to work with much bigger datasets (100s of GB)
```
you are able

joelostblom commented:
Thanks @QuLogic ! Merging!

(Btw, do we have any rules or norms for merging our own pull requests? I keep doing it assuming that it is ok after having addressed all the feedback. Let me know if I should wait for one of you guys to merge instead)
